### PR TITLE
Classification implementation

### PR DESCRIPTION
**Pull request by ppdebreuck**
This PR enables **classification** with MODNet. (+ some basic tests on classification)

### Changes:

- new `task_type` argument in MODData: optional list of integers, which has the same length as the number of targets (properties). The inner values of the list are set to  1 for a classification task and 0 for a regression task. If left to None, all tasks are considered regressions.
- new `task_type` argument in MODNetModel: dictionary defining the target types (classification or regression). Should be constructed as follows: key: string giving the target name; value: integer n, with n=0 for regression and n>=2 for classification with n the number of classes.
- MODData feature selection is based on the scikit-learn `mutual_info_classif` function
- MODNetModel construct n sigmoid neurons for each classification task, with n the number of classes.

### Important notes:
- For single or multi-target target classification, the user should use an appropriate loss functions (e.g. cross entropy loss).
- For mixed classification and regression: should work (not tested); but a good loss function should be used to combine both.

### TODO (other PR or simple commit):
- Update readme accordingly
- Implement a mixed classification and regression loss function (automatically set).

### CODE REVIEW

File: modnet/tests/test_preprocessing.py [hunk 1]
```diff
@@ -105,6 +105,70 @@ def test_nmi_target():
...
+    # in mutual_info_regression)
+    npoints = 1500
+    x = np.array([0]*500+[1]*500+[20]*500,dtype='float')
+    y = 2 * x
+    z = np.array(x,dtype='int')
+    print(z)
+
+    df_feat = pd.DataFrame({'x': x, 'y': y})
+    df_target = pd.DataFrame({'z': z})
+    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type=1, n_neighbors=2)
```
Comment (ml-evs) [inline modnet/tests/test_preprocessing.py:119]:
```suggestion
    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type="classification", n_neighbors=2)
```

File: modnet/tests/test_preprocessing.py [hunk 2]
```diff
@@ -105,6 +105,70 @@ def test_nmi_target():
...
+    np.random.seed(42)
+    np.random.shuffle(indices)
+    xs = x.take(indices)
+    ys = y.take(indices)
+    zs = z.take(indices)
+
+    df_feat = pd.DataFrame({'x': xs, 'y': ys})
+    df_target = pd.DataFrame({'z': zs})
+
+    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type=1, n_neighbors=2)
```
Comment (ml-evs) [inline modnet/tests/test_preprocessing.py:137]:
```suggestion
    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type="classification", n_neighbors=2)
```

File: modnet/tests/test_preprocessing.py [hunk 3]
```diff
@@ -105,6 +105,70 @@ def test_nmi_target():
...
+    assert df_nmi_target.shape == (2, 1)
+    assert df_nmi_target.loc['x']['z'] == pytest.approx(1.0,0.02)
+    assert df_nmi_target.loc['y']['z'] == pytest.approx(1.0,0.02)
+
+    # Test with one constant feature
+    c = np.ones(npoints) * 1.4
+    df_feat = pd.DataFrame({'x': x, 'y': y, 'c': c})
+    df_target = pd.DataFrame({'z': z})
+
+    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type=1, n_neighbors=2)
```
Comment (ml-evs) [inline modnet/tests/test_preprocessing.py:148]:
```suggestion
    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type="classification", n_neighbors=2)
```

File: modnet/tests/test_preprocessing.py [hunk 4]
```diff
@@ -105,6 +105,70 @@ def test_nmi_target():
...
+    assert df_nmi_target.shape == (2, 1)
+    assert df_nmi_target.loc['x']['z'] == pytest.approx(1.0, 0.02)
+    assert df_nmi_target.loc['y']['z'] == pytest.approx(1.0, 0.02)
+
+    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type=1, drop_constant_features=False, n_neighbors=2)
```
Comment (ml-evs) [inline modnet/tests/test_preprocessing.py:153]:
```suggestion
    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type="classification", drop_constant_features=False, n_neighbors=2)
```

File: modnet/tests/test_preprocessing.py [hunk 5]
```diff
@@ -105,6 +105,70 @@ def test_nmi_target():
...
+    x = np.linspace(start=2, stop=5, num=4)
+    z = np.linspace(start=3, stop=7, num=5)
+    x, z = np.meshgrid(x, z)
+    x = x.flatten()
+    z = z.flatten()
+    z = np.array(z / 10, dtype='int')
+    df_feat = pd.DataFrame({'x': x})
+    df_target = pd.DataFrame({'z': z})
+
+    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type=1)
```
Comment (ml-evs) [inline modnet/tests/test_preprocessing.py:168]:
```suggestion
    df_nmi_target = nmi_target(df_feat=df_feat, df_target=df_target, task_type="classification")
```

File: modnet/models.py [hunk 6]
```diff
@@ -46,6 +47,9 @@ def __init__(
             targets: A nested list of targets names that defines the hierarchy
                 of the output layers.
             weights: The relative loss weights to apply for each target.
+            task_type: Dictionary defining the target types (classification or regression).
+                Should be constructed as follows: key: string giving the target name; value: integer n,
+                 with n=0 for regression and n>=2 for classification with n the number of classes.
```
Comment (ml-evs) [inline modnet/models.py:52]: I like this more than corresponding `MODData` version, but perhaps a different name is better? `num_classes`?

**Review by ml-evs — COMMENTED** [review]
Thanks for this @ppdebreuck! 

I've made a few specific comments/suggestions below.

I have a couple of general queries:

- Should `MODData` do some processing of the targets so that e.g. they are all mapped to integers? Or should it at least warn if they can't be/aren't? In this case the classes could be easily counted in MODData, so `MODNetModel` would not need to know about the number, just that they should be treated as classification tasks.
- Should `model.fit(...)` check that the `MODData` has been prepared for the appropriate task too?

File: modnet/preprocessing.py [hunk 7]
```diff
@@ -79,14 +80,25 @@ def nmi_target(df_feat: pd.DataFrame, df_target: pd.DataFrame,
         to_drop = frange[frange == 0].index
         df_feat = df_feat.drop(to_drop, axis=1)
 
+    # Take right MI fun depending on classif/regression
+    if task_type: # Classification
+        _mifun = mutual_info_classif
+    else: # Regression
+        _mifun = mutual_info_regression
```
[the remark this answers is not in the document]
Comment (ppdebreuck) [reply modnet/preprocessing.py:87]: Great thanks for the 'partial' suggestion! But, `discrete_features` should be False all the time except when computing target-target NMI for normalizing, so should probably change this.

File: modnet/preprocessing.py [hunk 8]
```diff
@@ -79,14 +80,25 @@ def nmi_target(df_feat: pd.DataFrame, df_target: pd.DataFrame,
...
 
     # Compute the "self" mutual information (i.e. information entropy) of the target variable and of the input features
-    target_mi = mutual_info_regression(df_target[target_name].values.reshape(-1, 1),
-                                       df_target[target_name], **kwargs)[0]
+    if task_type:
+        target_mi = _mifun(df_target[target_name].values.reshape(-1, 1),
+                                       df_target[target_name], discrete_features=True, **kwargs)[0]
+    else:
+        target_mi = _mifun(df_target[target_name].values.reshape(-1, 1),
+                           df_target[target_name], **kwargs)[0]
```
[the remark this answers is not in the document]
Comment (ppdebreuck) [reply modnet/preprocessing.py:101]: If classification: this is the only call that needs to have 'discrete_features=True', all others discrete_features=False or 'auto'.

**Comment by ppdebreuck** [comment]
Thanks @ml-evs for the suggestions! Will make changes accordingly. For your general comments:
- I am considering that the data is already mapped to integers starting from 0. We should probably warn or specify this.
- The problem is that MODNetModel is constructed without moddata (it is only proved later in the fit method), so without any guess on the type of class, there is no way to build the model. One idea is to include a moddata while building the model (on the object instantiation), or merge the init and fit function single build+fit function).

File: modnet/models.py [hunk 9]
```diff
@@ -201,7 +220,14 @@ def fit(
         x = training_data.get_featurized_df()[
             self.optimal_descriptors[:self.n_feat]
         ].values
-        y = [training_data.df_targets[targ].values.astype(np.float, copy=False) for targ in self.targets_flatten]
+
```
Comment (ml-evs) [inline modnet/models.py:223]: Could we also default to a better loss when doing a classification here (KL or otherwise)?
